### PR TITLE
nzxt-kraken3: PWM setting fixes; status reporting optimization

### PR DESCRIPTION
**Pull request by aleksamagicka**
This should resolve the issues noted in #7 and optimize Z series status reporting.

- Values that convert to less than `PUMP_DUTY_MIN` will now be silently set to it without returning error codes
- When setting PWM duty directly, the same value that was set will be reported until the next sensor report interrupt comes through, so that userspace tools will be happ(y/ier)
- I've removed `priv->duty_input` in favor of recording the value in `kraken3_channel_info` for each channel
- Z series don't send the sensor report on their own. Instead of requesting it every time something is read, request it only when the validity interval has passed
- Reformatted code

---

Seems to work when using `fan2go`, here's my config: [fan2go.txt](https://github.com/liquidctl/liquidtux/files/10369233/fan2go.txt). I had to set `maxRpmDiffForSettledFan: 20` because X53 varied between 14-20 as far as I've seen, and the default value is 10. Didn't yet test with pwmconfig and fancontrol.

### CODE REVIEW

**Comment by stalkerg** [comment]
Thanks a lot! I will check today or tomorrow!

File: nzxt-kraken3.c [hunk 1]
```diff
@@ -333,6 +345,14 @@ static int kraken3_write(struct device *dev, enum hwmon_sensor_types type, u32 a
...
 				if (ret < 0)
 					return ret;
+
+				/*
+				 * Lock onto this value and report it until next interrupt status
+				 * report is received, so userspace tools can continue to work
+				 */
+				spin_lock_bh(&pwm_write_lock);
+				priv->channel_info[channel].reported_duty = val;
+				spin_unlock_bh(&pwm_write_lock);
```
[the remark this answers is not in the document]
Comment (aleksamagicka) [reply nzxt-kraken3.c:355]: Hm, might be. Will investigate later, this was a first shot at resolving this

[the remark this answers is not in the document]
Comment (stalkerg) [reply nzxt-kraken3.c:355]: yes, `reported_duty` is u16 it should be trivial to do as atomic.

**Comment by stalkerg** [comment]
It seems like working as expected, I also have no any significant issues with fan2go. Thanks again!

File: nzxt-kraken3.c [hunk 2]
```diff
@@ -197,9 +207,9 @@ static int kraken3_pwm_to_percent(long val, int channel)
 
 	percent_value = DIV_ROUND_CLOSEST(val * 100, 255);
 
-	/* Pump has a minimum duty value in addition to the max */
-	if ((channel == 0 && percent_value < PUMP_DUTY_MIN) || percent_value > CURVE_DUTY_MAX)
-		return -EINVAL;
+	/* Bring up pump duty to min value if needed */
+	if (channel == 0 && percent_value < PUMP_DUTY_MIN)
+		percent_value = PUMP_DUTY_MIN;
```
Comment (jonasmalacofilho) [inline nzxt-kraken3.c:212]: Shouldn't `-EINVAL` still be returned if `percent_value > CURVE_DUTY_MAX`?

File: nzxt-kraken3.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonasmalacofilho) [reply nzxt-kraken3.c:355]: I also don't understand this spinlock: unless it's also taken when reading, it doesn't do much.

---

@amezin ,

Regarding 16-bit stores being atomic: they aren't, at least not in the context of either the C or the Linux memory models (well, the one documented in [`tools/memory-model/Documentation/`](https://github.com/torvalds/linux/blob/1fe4fd6f5cad/tools/memory-model/Documentation/explanation.txt#L2044-L2057)). By those definitions, either store to `reported_duty` can cause a data race with each other or the load in `kraken3_read`.

(For completeness: in the context of the CPU, 16-bit stores _might_ be compiled into a single/simple store, in which case they _would_ be atomic _on x86._ Same thing for the load, and looking at the surrounding code, this is probably the case here. And I suppose there is almost no one connecting these coolers to an ARM machine, even if just for external monitoring of a x86 system).

But it seems that, at least in the HWMON subsystem (and, to be fair, other parts of the kernel), plain accesses can be used even if they _theoretically_ can cause data races, as long as the compiler (currently) produces sensible code (reminder: a data race is UB), and that the impact of a perhaps stale or garbage value isn't relevant or can't be observed in practice.

I say this based on a discussion I had with Guenter on the mailing list. There's a bit of noise due to some communication issues, but you can find the entire thread [here](https://lore.kernel.org/linux-hwmon/20210329082211.86716-1-jonas@protocubo.io/).

TL;DR: @aleksamagicka, removing the spinlock and just keeping the plain store and load is probably fine. But please test it well.

(Foreshadowing: this is fine meme).

---

@stalkerg,

Yes, _making_ the store and load atomic, or even just "marked" (i.e. using the kernel's `READ_ONCE`/`WRITE_ONCE` macros, which essentially do volatile accesses), would also be possible.

Still, I think the (usually) cheaper plain accesses are preferred if no issues can be observed when using them.

File: nzxt-kraken3.c (hunk 2 above)
Comment (aleksamagicka) [reply nzxt-kraken3.c:212]: Unless I'm mistaken it can't be. Max PWM is 255, so max `percent_value` is 100.

File: nzxt-kraken3.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (aleksamagicka) [reply nzxt-kraken3.c:355]: The spinlock was me copying the nzxt-smart code, thinking about it it probably does not do anything here, I'll test.

File: nzxt-kraken3.c (hunk 2 above)
Comment (aleksamagicka) [reply nzxt-kraken3.c:212]: It probably doesn't, but [we do](https://github.com/liquidctl/liquidtux/blob/7d6ce4d3934b882e15d21b2bd2bebfbdf464e693/nzxt-kraken3.c#L205).

Comment (jonasmalacofilho) [reply nzxt-kraken3.c:212]: Oh, right, totally missed it even though it was right there, 3 lines above :facepalm:

File: nzxt-kraken3.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (stalkerg) [reply nzxt-kraken3.c:355]: Sure, it seems like here we can just remove spinlock but if we need atomic, I would prefer to use `atomic_set`. 
https://www.kernel.org/doc/html/v4.15/core-api/atomic_ops.html

[the remark this answers is not in the document]
Comment (jonasmalacofilho) [reply nzxt-kraken3.c:355]: @stalkerg,

We don't need atomic per say. Atomics accesses are one way to prevent data races, but they are costly on some architectures, [^1] and there are other ways too (locks, for example, plus the marked accesses I mentioned before).

More importantly, I think what the subsystem maintainers (i.e. Guenter) prefers is what matters here. Then if future compiler versions start breaking hwmon code, it would be up to him to push back (or change the code as needed).

[^1]: Although atomic loads with acquire ordering and atomic stores with release ordering are free on x86. Well, more accurately, they're already always paid for.

[the remark this answers is not in the document]
Comment (aleksamagicka) [reply nzxt-kraken3.c]: Removed the spinlock after testing a few times, seems to work fine. Will test a few more times in the upcoming days.

The spinlock might not have done much locking, but as the function names suggest it does disable bottom half routines - it remains to be seen/tested if that applies to kraken3_raw_event() as the actual processing function for interrupts. But, judging from Guenter's emails (thanks for the link!), it's not terribly important for this to be synchronized because it's for display/info only. (Do correct me if I got that wrong.)

[the remark this answers is not in the document]
Comment (jonasmalacofilho) [reply nzxt-kraken3.c]: > But, judging from Guenter's emails (thanks for the link!), it's not terribly important for this to be synchronized because it's for display/info only. (Do correct me if I got that wrong.)

That's more or less what I understood from his comments too.

> The spinlock might not have done much locking, but as the function names suggest it does disable bottom half routines - it remains to be seen/tested if that applies to kraken3_raw_event() as the actual processing function for interrupts.

In the past I remember looking into on which context the raw event handler runs.

Assuming this [old comment of mine](https://github.com/liquidctl/liquidtux/blob/b3e18016e09b9a7077e5d23988af102f6376899a/krx62.c#L147) is correct, the handler runs in a hard IRQ. And IIRC, that means that `spin_lock_irq[save]` would be necessary to disable it.

[the remark this answers is not in the document]
Comment (stalkerg) [reply nzxt-kraken3.c]: @jonasmalacofilho I know, I worked with atomic a lot, especially for PowerPC. Just in case, spinlock internally works based on an atomic operation (CAS or TAS). It means if possible using atomics directly it will not be so costly as spinlock in any case. But as Guenter is mention - in such a case we shouldn't worry about guarding `reported_duty` at all.

[the remark this answers is not in the document]
Comment (amezin) [reply nzxt-kraken3.c]: @jonasmalacofilho Ok, I should've said "effectively atomic". I. e. another thread will read either the old value or new one, not garbage. Even on ARM. Not in generic C, but in the Linux kernel with supported compilers and flags. Or is this incorrect too?

**Comment by amezin** [comment]
@aleksamagicka @stalkerg By the way, have you tested the driver with lm-sensors (i. e. pwmconfig+fancontrol)? These utilities sometimes fail in surprising ways. As far as I remember, they treat almost any sysfs read error as "oh the device is completely broken let's die".

**Comment by stalkerg** [comment]
@amezin I didn't test it in fancontrol bu fan2go has totally the same behavior for read error (and it's ok).

File: nzxt-kraken3.c [hunk 3]
```diff
@@ -210,21 +213,23 @@ static int kraken3_read(struct device *dev, enum hwmon_sensor_types type, u32 at
...
+			/* Send command for getting status */
+			ret = kraken3_write_expanded(priv, z53_get_status_cmd,
+						     Z53_GET_STATUS_CMD_LENGTH);
+			if (ret < 0)
+				return ret;
 
-		wait_for_completion(&priv->z53_status_processed);
+			wait_for_completion(&priv->z53_status_processed);
+		} else {
+			return -ENODATA;
```
Comment (amezin) [inline nzxt-kraken3.c:229]: Btw, nzxt-smart2 doesn't check last updated time. Just always returns the latest data. Are there any reasons for the check? Should I implement it in my driver maybe?

As I've said, read errors usually make fancontrol shut down completely (if I remember correctly).

File: nzxt-kraken3.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (jonasmalacofilho) [reply nzxt-kraken3.c]: @amezin ,

> Not in generic C, but in the Linux kernel with supported compilers and flags. Or is this incorrect too?

According to the LKMM:[^1]

> if it [the program] does [contain race candidates which may execute concurrently] then the LKMM says there is a potential data race and makes no predictions about the program's outcome.

Regarding compiler flags used by the kernel to make compilers work more conservatively, I'm only aware of a few of them, related to other types of optimizations. So I can't answer that part.

By the way, there's an interesting (and reasonably recent) article on LWM about this: [Who's afraid of a big bad optimizing compiler?](https://lwn.net/Articles/793253/)

P.S. I have the impression that there are two groups within the kernel: some tend to approach these problems from the point of view of what the compiler is allowed to do; some from the point of view of whether stricter accesses or synchronization can be shown to be necessary.

P.P.S. Personally, I find performance arguments _against_ `READ_ONCE`/`WRITE_ONCE` not compelling when they are a viable way to prevent a clear potential data race: if those marked accesses prevent optimizations, that usually suggests that they are indeed necessary; and if they don't, they are free now, and a safeguard against future compiler versions producing different and potentially broken code.

[^1]: https://github.com/torvalds/linux/blob/1fe4fd6f5cad/tools/memory-model/Documentation/explanation.txt#L2065-L2068

File: nzxt-kraken3.c (hunk 3 above)
Comment (jonasmalacofilho) [reply nzxt-kraken3.c:229]: Regarding fancontrol, I think that its approach to error tolerance – no tolerance whatsoever, just kill the process – is unreasonable, especially for a fan control daemon. I imagine most users don't even set up alerts if the service dies.

Comment (amezin) [reply nzxt-kraken3.c:229]: Created a reminder for myself: https://github.com/liquidctl/liquidtux/issues/35

File: nzxt-kraken3.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (aleksamagicka) [reply nzxt-kraken3.c]: @jonasmalacofilho Thanks, if it runs in hard IRQ then those spinlock bottom half calls really do nothing.
